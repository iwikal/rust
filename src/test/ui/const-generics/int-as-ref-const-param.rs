// normalize-stderr-32bit: "0x" -> "$$PREFIX"
// normalize-stderr-64bit: "0x00000000" -> "$$PREFIX"

#![feature(const_generics, const_compare_raw_pointers)]
//~^ WARN the feature `const_generics` is incomplete and may cause the compiler to crash

struct Const<const P: &'static ()>;

#[repr(C)]
union Transmuter {
    pointer: *const (),
    reference: &'static (),
}

fn main() {
    const A: &'static () = {
        unsafe { Transmuter { pointer: 10 as *const () }.reference }
    };
    const B: &'static () = {
        unsafe { Transmuter { pointer: 11 as *const () }.reference }
    };

    let _: Const<{A}> = Const::<{B}>; //~ mismatched types
    let _: Const<{A}> = Const::<{&()}>; //~ mismatched types
    let _: Const<{A}> = Const::<{A}>;
}
