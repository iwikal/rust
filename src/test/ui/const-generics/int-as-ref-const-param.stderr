warning: the feature `const_generics` is incomplete and may cause the compiler to crash
  --> $DIR/int-as-ref-const-param.rs:4:12
   |
LL | #![feature(const_generics, const_compare_raw_pointers)]
   |            ^^^^^^^^^^^^^^
   |
   = note: `#[warn(incomplete_features)]` on by default

error[E0308]: mismatched types
  --> $DIR/int-as-ref-const-param.rs:23:25
   |
LL |     let _: Const<{A}> = Const::<{B}>;
   |                         ^^^^^^^^^^^^ expected `$PREFIX0000000a : &()`, found `$PREFIX0000000b : &()`
   |
   = note: expected type `Const<$PREFIX0000000a : &()>`
              found type `Const<$PREFIX0000000b : &()>`

error[E0308]: mismatched types
  --> $DIR/int-as-ref-const-param.rs:24:25
   |
LL |     let _: Const<{A}> = Const::<{&()}>;
   |                         ^^^^^^^^^^^^^^ expected `$PREFIX0000000a : &()`, found `{pointer} : &()`
   |
   = note: expected type `Const<$PREFIX0000000a : &()>`
              found type `Const<{pointer} : &()>`

error: aborting due to 2 previous errors

For more information about this error, try `rustc --explain E0308`.
